hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java

```diff
@@ -134,6 +147,30 @@
         return plans;
     }
 
+    private static Address[] initPartitionOwnersAndMembers(NodeEngine nodeEngine, Collection<Member> members) {
+        Set<Address> addresses = members.stream().map(Member::getAddress).collect(toSet());
+        ClusterService clusterService = nodeEngine.getClusterService();
+
+        IPartitionService partitionService = nodeEngine.getPartitionService();
+        Address[] partitionOwners = new Address[partitionService.getPartitionCount()];
+        for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
+            Address address = partitionService.getPartitionOwnerOrWait(partitionId);
+            partitionOwners[partitionId] = address;
+
+            if (!addresses.contains(address)) {
+                // Address in partition table doesn't exist in initial member list.
+                // Check whether it's a newly joined member?
+                Member member = clusterService.getMember(address);
+                if (member == null) {
+                    throw new TopologyChangedException("Topology changed! " + address + " is not member anymore!");
```

Comment:
This logic feels racy to me. What if the member leaves the cluster just after we have made this check? The next operation will fail anyway since it will try to invoke an operation on a member which is not part of the cluster.

If the member wasn't in the initial member list but there is a partition assigned to it - then it must mean it's a new member. This case will only be hit if the new member already leaves the cluster between the call where we got the initial member list and the call where we ask for the owner of the partition. Is there a way we can get a copy of the partition table atomically, and use it as the list of members as well and not use two separate lists which need reconciling?